ksp_plugin/pile_up.cpp

```diff
@@ -419,44 +426,153 @@ void PileUp::DeformPileUpIfNeeded(Instant const& t) {
+
+  auto const L̂_apparent = NormalizeOrZero(apparent_angular_momentum);
+  auto const L̂_actual = NormalizeOrZero(angular_momentum_);
+  // NOTE(egg):
+  // — The two sides of this disjunction are always equal;
+  // — technically this will also be true on the essential singularity.
+  bool const on_removable_singularity =
+      apparent_correction_axis == Bivector<double, ApparentPileUp>{} ||
+      actual_correction_axis == Bivector<double, NonRotatingPileUp>{};
+  bool const on_essential_singularity =
+      L̂_apparent == Bivector<double, ApparentPileUp>{} ||
+      L̂_actual == Bivector<double, NonRotatingPileUp>{};
+
+  bool const trivial_rotations = !correct_orientation ||
+                                 on_removable_singularity ||
+                                 on_essential_singularity;
+
+  Rotation<ApparentPileUp, EquivalentRigidPileUp>
+      apparent_pile_up_equivalent_rotation =
+          trivial_rotations
+              ? Rotation<ApparentPileUp, EquivalentRigidPileUp>::Identity()
+              : Rotation<ApparentPileUp, EquivalentRigidPileUp>(
+                    apparent_correction_axis,
+                    L̂_apparent,
+                    Commutator(apparent_correction_axis, L̂_apparent));
+  Rotation<NonRotatingPileUp, EquivalentRigidPileUp>
+      actual_pile_up_equivalent_rotation =
+          trivial_rotations
+              ? Rotation<NonRotatingPileUp, EquivalentRigidPileUp>::Identity()
+              : Rotation<NonRotatingPileUp, EquivalentRigidPileUp>(
+                    actual_correction_axis,
+                    L̂_actual,
+                    Commutator(actual_correction_axis, L̂_actual));
+  Rotation<ApparentPileUp, NonRotatingPileUp> const
+      tentative_attitude_correction =
+          actual_pile_up_equivalent_rotation.Inverse() *
+          apparent_pile_up_equivalent_rotation;
+
   // The angular velocity of a rigid body with the inertia and angular momentum
   // of the apparent parts.
   auto const apparent_equivalent_angular_velocity =
       apparent_angular_momentum / inertia_tensor;
   // The angular velocity of a rigid body with the inertia of the apparent
   // parts, and the angular momentum of the pile up.
-  auto const actual_equivalent_angular_velocity =
-      angular_momentum_ /
-      Identity<ApparentPileUp, NonRotatingPileUp>()(inertia_tensor);
+  auto actual_equivalent_angular_velocity =
+      angular_momentum_ / tentative_attitude_correction(inertia_tensor);
+
+  // So far we have only dealt with the essential singularity by removing it.
+  // We now need to deal with its neighbourhood, by ensuring that the tentative
+  // correction is not too big compared to the angular velocities involved.
+  Quaternion const q = tentative_attitude_correction.quaternion();
+  // α is the angle of the tentative attitude correction.
+  Angle const α =
+      2 * quantities::ArcTan(q.imaginary_part().Norm(), q.real_part());
```

Comment:
Yuck.  We need a function to extract the angle (and while we are at it, the axis) of a rotation (and a quaternion since that's the same thing).  Please open an issue.

Reply:
#2534.